Mathlib/NumberTheory/Padics/Harmonic.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2023 Koundinya Vajjha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Koundinya Vajjha
+-/
+
+import Mathlib.NumberTheory.Padics.PadicIntegers
+import Mathlib.Data.Int.Log
+/-!
+
+The nth Harmonic number is not an integer. We formalize the proof using
+2-adic valuations. This proof is due to Kürschák.
+
+Reference:
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/padicharmonicsum.pdf
+
+-/
+
+open BigOperators
+
+/-- The nth-harmonic number defined as a finset sum of consecutive reciprocals. -/
+def harmonic : ℕ → ℚ := fun n => ∑ i in Finset.range n, 1 / (i + 1)
+
+lemma harmonic_pos {n : ℕ} (Hn : n ≠ 0) : 0 < harmonic n :=
+  Finset.sum_pos (fun _ _ => div_pos zero_lt_one (by norm_cast; linarith))
```

Comment:
(And switching to `n⁻¹` should in theory make some proofs a bit tidier)
```suggestion
  Finset.sum_pos (fun _ _ => inv_pos.mpr (by norm_cast; linarith))
```

Reply:
It also breaks other proofs down below which I'm quite weary of fixing... If this is a matter of taste, I'd prefer to leave it as is. Other places in mathlib do use `1/x` as I've linked to above.

Reply:
I opened a discussion on Zulip: https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.237319.20harmonic.20number.20not.20an.20integer

Reply:
OK, Made the change. It wasn't that bad, actually.